File: 044_Wildcard_Matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        ns, np = len(s), len(p)
        match = [[False] * (np + 1) for _ in range(ns + 1)]
        match[ns][np] = True
        for i in range(np - 1, -1, -1):
            if p[i] != '*':
                break
            else:
                match[ns][i] = True
        for i in range(ns - 1, -1, -1):
            for j in range(np - 1, -1, -1):
                if s[i] == p[j] or p[j] == '?':
                    match[i][j] = match[i + 1][j + 1]
                elif p[j] == '*':
                    match[i][j] = match[i + 1][j] | match[i][j + 1]
                else:
                    match[i][j] = False
        return match[0][0]

File: 044_Wildcard_Matching.cpp
class Solution {
public:
    bool isMatch(string s, string p) {
        int i = 0, j = 0, star = -1, match = 0;
        while (i < s.size()) {
            // advance both iterator for both string and pattern
            if (j < p.size() && (s[i] == p[j] || p[j] == '?')) {
                ++i;
                ++j;
            }
            // advance only pattern
            else if (j < p.size() && p[j] == '*') {
                star = j++;
                match = i;
            }
            // advance only string
            else if (star > -1) {
                j = star + 1;
                i = ++match;
            }
            else return false;
        }
        for (; p[j] == '*'; ++j);
        return j == p.size();
    }
};
